wasm_runtime/src/c_api.rs

```diff
@@ -388,19 +397,33 @@ pub extern "C" fn wasm_executionctx_stdin_set(executionctx_id: *const c_char, bu
 /// wasm_return_const_char_ownership(module_output);
 /// ```
 #[no_mangle]
-pub extern "C" fn wasm_executionctx_run(executionctx_id: *const c_char) -> *const c_char {
+pub extern "C" fn wasm_executionctx_run(executionctx_id: *const c_char, _buffer: &mut *const c_char, _len: &mut c_ulong) -> c_int {
     let executionctx_id_str = const_c_char_to_str(executionctx_id);
 
-    let result = match WasmExecutionCtx::run(executionctx_id_str) {
-        Ok(output) => output,
+    let result: c_int = match WasmExecutionCtx::run(executionctx_id_str) {
+        Ok(output) => {
+            // 1) extract output length and try casting into `c_ulong` since it might not fit
+            *_len = match c_ulong::try_from(output.len()) {
+                Ok(l) => l,
+                Err(e) => {
+                    let error_msg = format!("ERROR! C-API: Couldn't cast Wasm module's output length into a C `unsigned long int` type! \'{}\' {:?}", output.len(), e);
+                    eprintln!("{}", error_msg); 
+                    0
```

Comment:
That's a good question.

Such conversion from `usize` to `c_ulong` should be always successful unless the `output.len()` is a really HUGE value that cannot fit into `c_ulong`. But `_len` cannot get a `-1` since it cannot hold negative numbers. 

I see two options here:
1) Leave as is, so even if the conversion of the `_len` fails and it's returned as `0`, the caller still can try to read the buffer for debugging purposes.
2) Instead of `-1` (which is not possible due to `_len` type), call directly to `return -1;` to finish the function and not return the output buffer.

Reply:
I see. Yeah, it's a weird error as `c_ulong` should be big enough. If the buffer is bigger than that, you may have other errors 😅.

> 1. Leave as is, so even if the conversion of the _len fails and it's returned as 0, the caller still can try to read the buffer for debugging purposes.
> 2. Instead of -1 (which is not possible due to _len type), call directly to return -1; to finish the function and not return the output buffer.

If we go for the first approach, users won't see any error unless they change the code to read the buffer even when the size is `0`. I mean, there's no flag or configuration option that modifies this behavior. I'm worried this may cause a "silent" error which may be more difficult to debug and even a memory leak.

Since this is not something trivial, I would return `-1` directly and let the function fails. Then, we will get a clear error and we can start debugging from that point.

Wdyt?

Reply:
Agree!